lib/rust/ensogl/core/src/display/object.rs

```diff
@@ -23,3 +28,138 @@ pub mod traits {
     pub use super::Object as TRAIT_Object;
     pub use super::ObjectOps as TRAIT_ObjectOps;
 }
+
+
+
+// =============================
+// === InstanceWithSublayers ===
+// =============================
+
+/// A wrapper around [`Instance`] with attached sublayers. It is used for
+/// dynamic creation and managing of additional on-demand sublayers needed for various UI
+/// components.
+///
+/// When the Instance is added to another layer, [`Sublayers::remove_from_layer`] and
+/// [`Sublayers::add_to_layer`] methods of the `sublayers` field will be called with previous and
+/// new layers respectively.
+///
+/// See [`MaskingSublayers`] docs for one of the use cases.
+#[derive(Debug, Clone, CloneRef)]
+pub struct InstanceWithSublayers<S: CloneRef> {
+    inner:         Instance,
+    /// Attached sublayers. Should implement [`Sublayers`] trait.
+    pub sublayers: S,
+}
```

Comment:
Never, ever use trait bounds on struct parameters. I was writing about it a lot of times – this is antipattern, as you have to then use this bound everywhere, on every trait impl and it leaks everywhere. Use bounds only in impls when necessarily. The only exception are trait bounds enforced by the drop impl.

Reply:
This requires manual implementation of `Clone` and `CloneRef`, but yes, I agree with you.